### PR TITLE
Refactoring and simplifying the code for the banded diagonal matrix and solve system (Issue #323)

### PR DESCRIPTION
**Pull request by negin513**
### Description of changes

Previously, the SoilTemperature module has a high level of nested subroutine calls, which hurt its performance and readability. There were also many redundancies and logical issues in the modular code (For example https://github.com/ESCOMP/CTSM/issues/323#issuecomment-598034948).  The modular code had a high level of nested conditionals, which were not necessary.  Based on #301 , @swensosc   found that by completely inlining all of the calls to set up the matrix, the performance of the matrix construction is improved by 3X. 

In this PR, I tried to simplify and optimize this code as much as possible while maintaining bfb compatibility and keeping a lower level of modularity. 

For simplifying the matrix construction code, we discussed a few different possibilities. The discussion is documented [here](https://github.com/ESCOMP/CTSM/issues/323). 

Here I merged the `SetMatrix***` and `SetRHSVec***` for all columns (urban road, urban non-road, non-urban) into one subroutine for each layer type (i.e. soil, snow, standing surface water). 
For example, I combined all five subroutine: `SetMatrix_Snow` , `SetMatrix_SnowUrban`, `SetMatrix_SnowUrbanRoad`, `SetMatrix_SnowUrbanNonroad`, and `SetMatrix_ nonUrban` into one `SetMatrix_Snow`.  

This simplified everything drastically, so currently, we have:

*  SetMatrix
     * SetMatrix_Snow
     * SetMatrix_Soil
     * SetMatrix_StandingSurfaceWater

* SetRHSVec
    * SetRHSVec_Snow
    * SetRHSVec_Soil
    * SetRHSVec_StandingSurfaceWater

The subroutines for the interaction layers are folded inside the above subroutines as well.

This PR improved the performance of `SetMatrix` subroutine by **2X (100%)** and `SetRHSVec`  by **50%**.  
### Specific notes

Contributors other than yourself, if any: @swensosc 

CTSM Issues Fixed (include github issue #): #323

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? None

Testing performed, if any:
Tested `aux_clm` on Cheyenne. I will run standard testing ( `aux_clm`) for Izumi.

### CODE REVIEW

**Comment by negin513** [comment]
I would like to ask for reviewers'  comments on a few points which can help further optimize this code. 

1) In the original code, we have several nested conditionals like the following 
https://github.com/ESCOMP/CTSM/blob/c5a9030f57f3db13a23721af7bb6c4dd8ba00d84/src/biogeophys/SoilTemperatureMod.F90#L2133-L2137

I was wondering if it is possible for `col%itype(c`) to be equal to urban non road (either `icol_sunwall` or `icol_shadewall` or `icol_roof`) while `lun%urbpoi(l)` is not true.

If not,  `if (lun%urbpoi(l))` is redundant and we can remove this conditional to improve the performance. 

For this PR, I kept this if statement in my codes as following but I believe it is redundant and can be removed
https://github.com/ESCOMP/CTSM/blob/e4f0a50ded65cf01e2716039571b948acf1a4112/src/biogeophys/SoilTemperatureMod.F90#L2504-L2508

**Comment by dlawrenncar** [comment]
Hey Negin,

I suggest that you ping Keith Oleson directly (author of the urban code),
via github on this, just in case he is not tracking all CTSM github
messages.

Thanks for working on this.  Looks almost done!  Is the improved
performance detectable for the overall model or is it just detectable for
this routine?

Dave

**Comment by negin513** [comment]
2) For urban road columns we had the following in the original code:

https://github.com/ESCOMP/CTSM/blob/c5a9030f57f3db13a23721af7bb6c4dd8ba00d84/src/biogeophys/SoilTemperatureMod.F90#L4092-L4121


Here we have a `else if (j == 1)` (L4102) and then the next condition is `else if (j <= nlevgrnd-1)` (L4114). I think whenever `else if (j == 1)` is true   `else if (j <= nlevgrnd-1)` is true as well and everything will be overwritten in by `else if (j <= nlevgrnd-1)`....

This is besides the wrong code with logical issue in the same loop that I discussed [here](https://github.com/ESCOMP/CTSM/issues/323#issuecomment-598034948)

**Comment by olyson** [comment]
I don't think this is possible:

I was wondering if it is possible for col%itype(c) to be equal to urban non road (either icol_sunwall or icol_shadewall or icol_roof) while lun%urbpoi(l) is not true.

**Comment by olyson** [comment]
Only one else or else if block should be entered for each j, right?  So I don't see how everything will be overwritten by else if (j <= nlevgrnd-1).

**Comment by negin513** [comment]
> Only one else or else if block should be entered for each j, right? So I don't see how everything will be overwritten by else if (j <= nlevgrnd-1).

I think you are correct. The reason I am asking this is that this block of `else if (j == 1)` can be simplified to one line:

https://github.com/ESCOMP/CTSM/blob/c5a9030f57f3db13a23721af7bb6c4dd8ba00d84/src/biogeophys/SoilTemperatureMod.F90#L4102-L4113

can be written as one line in the block of `else if (j <= nlevgrnd-1)`:

```
bmatrix_soil(c,3,1) = 1._r8 + (1._r8-cnfac)*fact(c,j)*(tk(c,j)/dzp & 
     + frac_sno_eff(c) * tk(c,j-1)/dzm) & 
     - (1._r8 - frac_sno_eff(c))*fact(c,j)*dhsdT(c)
```

everything else is similar for j==1 and j<nlevgrnd-1

**Comment by negin513** [comment]
> I don't think this is possible:
> 
> I was wondering if it is possible for col%itype(c) to be equal to urban non road (either icol_sunwall or icol_shadewall or icol_roof) while lun%urbpoi(l) is not true.

Perfect! Then this would simplify things further...

**Comment by olyson** [comment]
Regarding the lun%urbpoi(l) check, the original code, before it was rewritten by someone to do this extreme modularization, was probably part of a larger loop that had an else if to treat non-urban points.  So I expect there are more instances of this, glad to see you are fixing all of this.

**Comment by billsacks** [comment]
@swensosc - I'm adding you as a reviewer here because I know you were involved in the initial stages of this - and you're the one who requested this in the first place. However, I'll leave it up to you as to whether you actually want to take the time to review the final product. If you don't want to or don't have time to review it, let us know and we can remove you as a reviewer.

**Comment by negin513** [comment]
I also am not sure if we ever need this conditional `if (j >= col%snl(c)+1)` inside our soil loops when j is looping over from 1 to `nlevgrnd` (or `nlevurb`). Is that correct?

For example here in the original code:

https://github.com/ESCOMP/CTSM/blob/c5a9030f57f3db13a23721af7bb6c4dd8ba00d84/src/biogeophys/SoilTemperatureMod.F90#L4088-L4130

**Comment by olyson** [comment]
I don't see any need for that either.

**Comment by negin513** [comment]
> Hey Negin, I suggest that you ping Keith Oleson directly (author of the urban code), via github on this, just in case he is not tracking all CTSM github messages. Thanks for working on this. Looks almost done! Is the improved performance detectable for the overall model or is it just detectable for this routine? Dave

@dlawrenncar This is a great question! From my quick observations, these two parts (SetMatrix and SetRHS...) of this module, take almost ~5-6% of total clm run time. Making it twice as fast will reduce the clm run time by ~3%. 
Once, this PR is reviewed and ready for merging, I will do some more testing to give you exact timing stats on this. But I guess overall in a CLM run, it is not much detectable (<5%).

**Comment by billsacks** [comment]
@negin513 I'm finally starting to review this - sorry for the delay. I'll take a look and add any review comments. It looks like @olyson has answered all of the questions you asked above; are there still any questions that you would like more input on?

One general thing I noticed in your last commit - and I'm not sure if this happened elsewhere as well - is that you deleted comments like `! at` and `! bt`. Would people (like @swensosc ) prefer that these comments be left in place to make clear the correspondence between lines of code and terms in the matrix / vectors?

**Comment by negin513** [comment]
> @negin513 I'm finally starting to review this - sorry for the delay. I'll take a look and add any review comments. It looks like @olyson has answered all of the questions you asked above; are there still any questions that you would like more input on?
> 
> One general thing I noticed in your last commit - and I'm not sure if this happened elsewhere as well - is that you deleted comments like `! at` and `! bt`. Would people (like @swensosc ) prefer that these comments be left in place to make clear the correspondence between lines of code and terms in the matrix / vectors?

@billsacks Thanks for reviewing this PR. I added `!at` , `!bt`, and `!ct` comments as I was going through this code and matching it with CLM4.5 codes. The older CLM4.5 codes used `at`, `bt`, and `ct` vectors , while the recent version only used a 3D `bmatrix` and each dimension of `bmatrix` represented one of these vectors, but they are not mentioned explicitly in the code as `at`, `bt`, and `ct`.

File: src/biogeophys/SoilTemperatureMod.F90 [hunk 1]
```diff
@@ -2439,1802 +2119,106 @@ subroutine SetRHSVec_Soil(bounds, num_nolakec, filter_nolakec, &
...
+      ! urban non-road columns ------------------------------------------------------------------
+      !
+      do j = 1,nlevurb
+        do fc = 1,num_nolakec
+           c = filter_nolakec(fc)
+           if (col%itype(c) == icol_sunwall  .or. &
+               col%itype(c) == icol_shadewall .or. &
+               col%itype(c) == icol_roof) then
+
+              if (j >= col%snl(c)+1) then
```
[the remark this answers is not in the document]
Comment (negin513) [reply src/biogeophys/SoilTemperatureMod.F90:2131]: This is another example of what was in the previous code which were logically redundant and caused poor performance and confusion. 
Here, I mentioned it for another loop. 
I will fix them
https://github.com/ESCOMP/CTSM/pull/979#issuecomment-612203289

**Comment by billsacks** [comment]
> @billsacks Thanks for reviewing this PR. I added `!at` , `!bt`, and `!ct` comments as I was going through this code and matching it with CLM4.5 codes. The older CLM4.5 codes used `at`, `bt`, and `ct` vectors , while the recent version only used a 3D `bmatrix` and each dimension of `bmatrix` represented one of these vectors, but they are not mentioned explicitly in the code as `at`, `bt`, and `ct`.

Ah, thanks for the clarification. So your recent commit was just removing the comments you had added. That's fine, then.

**Comment by negin513** [comment]
This branch is ready to be merged with master. I updated 'ChangeLog' and `ChangeSum`.

**Comment by billsacks** [comment]
It doesn't look like you have pushed the updated ChangeLog and ChangeSum.

**Comment by negin513** [comment]
Thanks Bill. I've updated `ChangeLog` and `ChangeSum` and pushed it back.
The failed test ( `SMS_D_Ld5_Vnuopc.f10_f10_musgs.I2000Clm50BgcCropGs.cheyenne_intel.clm-default`) works by updating the externals.

**Pull request merged**